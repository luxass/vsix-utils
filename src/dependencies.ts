import type { Manifest, PackageManager } from "./types";
import { exec } from "node:child_process";
import path, { isAbsolute, join, resolve } from "node:path";
import process from "node:process";
import { promisify } from "node:util";

const execAsync = promisify(exec);

export interface ExtensionDependenciesOptions {
  /**
   * The package manager to use.
   */
  packageManager: PackageManager;

  /**
   * The current working directory
   * @default process.cwd()
   */
  cwd: string;
}

export interface ExtensionDependency {
  /**
   * The name of the dependency.
   */
  name: string;

  /**
   * The version of the dependency.
   */
  version?: string;

  /**
   * The path to the dependency.
   */
  path: string;
}

interface PnpmDependency {
  from: string;
  version: string;
  resolved: string;
  path: string;
  dependencies: Record<string, PnpmDependency>;
}

interface YarnTreeNode {
  name: string;
  children: YarnTreeNode[];
}

interface YarnDependency {
  name: string;
  path: string;
  version: string;
  children: YarnDependency[];
}

/**
 * Retrieves all production dependencies for an extension based on the package manager being used.
 *
 * @param {Manifest} manifest - The extension's manifest object containing dependency information
 * @param {ExtensionDependenciesOptions} options - Configuration options for retrieving dependencies
 *
 * @returns {Promise<ExtensionDependenciesResult} Promise resolving to an object containing:
 * - dependencies: Array of extension dependencies with name, version and path
 * - packageManager: The package manager that was used
 *
 * @throws Error if:
 * - Package manager cannot be detected when using 'auto'
 * - Unsupported package manager is detected/specified (e.g. deno, bun)
 * - Unable to parse dependency information from package manager output
 *
 * @remarks
 * Supports npm, yarn, and pnpm package managers.
 * When using npm, parses output of `npm list`
 * When using yarn, parses output of `yarn list`
 * When using pnpm, parses output of `pnpm list`
 */
export async function getExtensionDependencies(manifest: Manifest, options: ExtensionDependenciesOptions): Promise<ExtensionDependency[]> {
  const {
    packageManager,
    cwd = process.cwd(),
  } = options;

  const dependencies = new Set<ExtensionDependency>();

  if (packageManager === "npm") {
    const { stdout } = await execAsync("npm list --production --parseable --depth=99999 --loglevel=error", { cwd });
    const lines = stdout.split(/[\r\n]/).filter((path) => isAbsolute(path));

    for (const line of lines) {
      if (line === resolve(cwd)) {
        continue;
      }

      const dependency = line.split(`${path.sep}node_modules${path.sep}`)[1];

      if (dependency == null) {
        throw new Error(`could not parse dependency: ${line}`);
      }

      dependencies.add({
        name: dependency.replace(/\\/g, "/"),
        version: manifest.dependencies != null ? manifest.dependencies[dependency.replace(/\\/g, "/")] : undefined,
        path: line,
      });
    }
  } else if (packageManager === "yarn") {
    const { stdout } = await execAsync("yarn list --prod --json", { cwd });

    const match = /^\{"type":"tree".*$/m.exec(stdout);

    if (!match || match.length !== 1) {
      throw new Error("Could not parse result of `yarn list --json`");
    }

    const trees = JSON.parse(match[0]).data.trees as YarnTreeNode[];

    if (!Array.isArray(trees) || trees.length === 0) {
      return [];
    }

    const prune = true; // TODO: using packaged dependencies

    function asYarnDependency(prefix: string, tree: YarnTreeNode, prune: boolean): YarnDependency | null {
      if (prune && /@[\^~]/.test(tree.name)) {
        return null;
      }

      let name: string;
      let version: string = "";
      try {
        const tmp = tree.name.split("@");

        if (tmp[0] === "") {
          tmp.shift();
          tmp[0] = `@${tmp[0]}`;
        }

        name = tmp[0]!;
        version = tmp[1] || "";
      } catch {
        name = tree.name.replace(/^([^@+])@.*$/, "$1");
      }

      const dependencyPath = path.join(prefix, name);
      const children: YarnDependency[] = [];

      for (const child of tree.children || []) {
        const dep = asYarnDependency(path.join(prefix, name, "node_modules"), child, prune);

        if (dep) {
          children.push(dep);
        }
      }

      return { name, path: dependencyPath, children, version };
    }

    const result = trees.map((tree) => asYarnDependency(join(cwd, "node_modules"), tree, prune)).filter((dep) => dep != null);

    const internalDeps = new Set<string>();

    const flatten = (dep: YarnDependency) => {
      if (internalDeps.has(dep.path)) {
        return;
      }

      dependencies.add({
        name: dep.name,
        version: dep.version,
        path: dep.path,
      });
      internalDeps.add(dep.path);

      if (dep.children) {
        for (const child of dep.children) {
          flatten(child);
        }
      }
    };

    for (const dep of result) {
      flatten(dep);
    }
  } else if (packageManager === "pnpm") {
    // use --ignore-workspace to avoid always including the workspace packages
    const { stdout } = await execAsync("pnpm list --production --json --depth=99999 --loglevel=error --ignore-workspace", { cwd });
    let entryList = [];
    try {
      entryList = JSON.parse(stdout);
    } catch {
      return [];
    }

    if (!Array.isArray(entryList) || entryList.length === 0) {
      return [];
    }

    const entry = entryList[0] as {
      dependencies?: Record<string, PnpmDependency>;
    };

    if (entry == null || typeof entry !== "object" || entry.dependencies == null || typeof entry.dependencies !== "object") {
      return [];
    }

    const internalDeps = new Set<string>();

    const flatten = (dep: PnpmDependency) => {
      if (internalDeps.has(dep.path)) {
        return;
      }

      dependencies.add({
        name: dep.from,
        version: dep.version,
        path: dep.path,
      });
      internalDeps.add(dep.path);

      if (dep.dependencies) {
        for (const child of Object.values(dep.dependencies)) {
          flatten(child);
        }
      }
    };

    for (const value of Object.values(entry.dependencies)) {
      flatten(value);
    }
  } else {
    throw new Error(`unsupported package manager: ${packageManager}`);
  }

  return Array.from(dependencies);
}
