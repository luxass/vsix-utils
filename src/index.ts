/**
 * @module
 *
 * This module contains utility functions for handling files.
 */

export type {
  CollectOptions,
  ContentTypeResult,
  ExtensionDependenciesOptions,
  ExtensionDependency,
  ProcessedFiles,
  VsixFile,
  VsixInMemoryFile,
  VsixLocalFile,
} from "./files";
export {
  collect,
  getContentTypesForFiles,
  getExtensionDependencies,
  getExtensionPackageManager,
  isInMemoryFile,
  isLocalFile,
  processFiles,
} from "./files";

export type {
  ManifestAsset,
  ManifestAssetType,
  ProjectManifest,
  VsixManifestOptions,
} from "./manifest";

export {
  createVsixManifest,
  getManifestTags,
  isWebKind,
  readProjectManifest,
  transformExtensionKind,
} from "./manifest";
export { prepublish, type PrepublishOptions } from "./scripts";
export type {
  Manifest,
  PackageManager,
  PackageManagerWithAuto,
} from "./types";
export type { ManifestValidation } from "./validation";
export {
  ALLOWED_SPONSOR_PROTOCOLS,
  EXTENSION_NAME_REGEX,
  EXTENSION_PRICING,
  GITHUB_BADGE_URL_REGEX,
  VALID_EXTENSION_KINDS,
  validateProjectManifest,
  validateVSCodeTypesCompatability,
  VSCODE_ENGINE_COMPATIBILITY_REGEX,
} from "./validation";

export type {
  RawVsixPackage,
  ReadVsixOptions,
  WriteVsixOptions,
} from "./zip";

export {
  readVsix,
  writeVsix,
} from "./zip";
