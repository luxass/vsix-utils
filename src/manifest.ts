/**
 * This module contains utility functions for manifests.
 * @module manifest
 *
 * @example
 * ```ts
 * import { readProjectManifest } from "vsix-utils";
 *
 * const projectManifest = await readProjectManifest('/path/to/project');
 *
 * if (projectManifest != null) {
 *   console.log("an error occurred while reading the project manifest");
 * }
 *
 * const { fileName, manifest } = projectManifest;
 *
 * console.log(fileName); // Outputs: /path/to/project/package.json
 * console.log(manifest); // Outputs: Parsed content of package.json
 * ```
 */

import type { Manifest } from "./types";
import { readFile } from "node:fs/promises";
import path from "node:path";

export interface ProjectManifest {
  fileName: string;
  manifest: Manifest;
}

/**
 * Reads the project manifest (package.json) from the specified project directory.
 *
 * @param {string} projectDir - The directory of the project where the package.json is located.
 * @returns {Promise<ProjectManifest | null>} A promise that resolves to an object containing the file name and the parsed manifest.
 *
 * @example
 * ```ts
 * const { fileName, manifest } = await readProjectManifest('/path/to/project');
 * console.log(fileName); // Outputs: /path/to/project/package.json
 * console.log(manifest); // Outputs: Parsed content of package.json
 * ```
 */
export async function readProjectManifest(projectDir: string): Promise<ProjectManifest | null> {
  try {
    const manifestPath = path.join(projectDir, "package.json");
    const manifest = JSON.parse(await readFile(manifestPath, "utf-8"));

    return {
      fileName: manifestPath,
      manifest,
    };
  } catch {
    return null;
  }
}
