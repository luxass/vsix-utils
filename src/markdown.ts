import type { Manifest } from "./types";
import { remark } from "remark";
import remarkTransformLinks from "remark-transform-links";

export interface TransformMarkdownOptions {
  /**
   * The markdown content to transform.
   */
  content: string;

  /**
   * Prepend all relative links in README.md with the specified URL.
   */
  baseContentUrl?: string;

  /**
   * Prepend all relative image links in README.md with the specified URL.
   */
  baseImagesUrl?: string;

  /**
   * Whether to rewrite the markdown content.
   * @default true
   */
  rewrite?: boolean;

  /**
   * The branch to use for the base URLs.
   * @default "HEAD"
   */
  branch?: string;
}

export async function transformMarkdown(manifest: Manifest, options: TransformMarkdownOptions): Promise<string> {
  const {
    content,
    rewrite = true,
    branch = "HEAD",
  } = options;

  if (!rewrite) {
    return content;
  }

  let { baseContentUrl, baseImagesUrl } = options;

  const baseUrls = inferBaseUrls(manifest, branch);

  if (baseUrls == null) {
    throw new Error("Couldn't detect the repository where this extension is published.");
  }

  if (baseContentUrl == null) {
    baseContentUrl = baseUrls.contentUrl;
  }

  if (baseImagesUrl == null) {
    baseImagesUrl = baseUrls.imagesUrl;
  }

  const detectedSyntax = detectMarkdownSyntax(content);

  const file = await remark()
    .data("settings", {
      bullet: detectedSyntax.bullet,
      bulletOrdered: detectedSyntax.bulletOrdered,
      emphasis: detectedSyntax.emphasis,
      listItemIndent: detectedSyntax.listItemIndent,
      fence: detectedSyntax.fence,
      strong: detectedSyntax.strong,
      rule: detectedSyntax.rule,
    })
    .use(remarkTransformLinks, {
      baseUrl(_, type) {
        if (type === "image" || type === "html_img" || type === "html_video") {
          return baseImagesUrl;
        }

        return baseContentUrl;
      },
    })
    .process(content);

  return String(file);
}

export interface InferredBaseUrls {
  contentUrl: string;
  imagesUrl: string;
}

export function inferBaseUrls(manifest: Manifest, branch?: string): InferredBaseUrls | null {
  let repository = null;

  if (typeof manifest.repository === "string") {
    repository = manifest.repository;
  } else if (manifest.repository && typeof manifest.repository === "object" && "url" in manifest.repository && typeof manifest.repository.url === "string") {
    repository = manifest.repository.url;
  }

  if (!repository) {
    return null;
  }

  // we expect this is a ssh url
  if (repository.startsWith("git@")) {
    repository = repository.replace(":", "/").replace("git@", "https://");
  }

  let url;

  try {
    url = new URL(repository);
  } catch {
    return null;
  }

  // slice at 1 to remove the leading slash and 3 to remove everything after the repo name
  const ownerWithRepo = url.pathname.split("/").slice(1, 3).join("/").replace(/\.git$/, "");

  const branchName = branch ?? "HEAD";

  if (url.hostname === "github.com") {
    return {
      contentUrl: `https://github.com/${ownerWithRepo}/blob/${branchName}`,
      imagesUrl: `https://github.com/${ownerWithRepo}/raw/${branchName}`,
    };
  }

  if (url.hostname === "gitlab.com") {
    return {
      contentUrl: `https://gitlab.com/${ownerWithRepo}/-/blob/${branchName}`,
      imagesUrl: `https://gitlab.com/${ownerWithRepo}/-/raw/${branchName}`,
    };
  }

  if (url.hostname === "gitea.com") {
    return {
      contentUrl: `https://gitea.com/${ownerWithRepo}/src/branch/${branchName}`,
      imagesUrl: `https://gitea.com/${ownerWithRepo}/raw/branch/${branchName}`,
    };
  }

  return null;
}

export interface DetectedMarkdownSyntax {
  bullet?: "-" | "+" | "*";
  bulletOrdered?: "." | ")";
  emphasis?: "_" | "*";
  strong?: "_" | "*";
  fence?: "`" | "~";
  listItemIndent?: "tab" | "one" | "mixed";
  rule?: "*" | "-" | "_";
}

/**
 * Analyzes markdown content and detects the syntax patterns used.
 *
 * @param {string} content - The markdown string to analyze
 * @returns {DetectedMarkdownSyntax} An object containing detected markdown syntax preferences with the following properties:
 * - `bullet`: The bullet style used for unordered lists ("-", "*", or "+")
 * - `bulletOrdered`: The marker style used for ordered lists ("." or ")")
 * - `emphasis`: The emphasis marker style ("*" or "_")
 * - `strong`: The strong emphasis marker style ("*" or "_")
 * - `fence`: The code fence style ("`" or "~")
 * - `listItemIndent`: The indentation style for list items ("tab", "one", "mixed")
 * - `rule`: The horizontal rule style ("*", "-", or "_")
 *
 * @example
 * ```ts
 * const markdown = `
 *  - List item
 *  - Another item
 * `;
 *
 * detectMarkdownSyntax(content) // Returns { bullet: "-", listItemIndent: "one" }
 * ```
 */
export function detectMarkdownSyntax(content: string): DetectedMarkdownSyntax {
  const result: DetectedMarkdownSyntax = {};

  const hyphenBullets = content.match(/^[\t ]*-[\t ][^\n]+/gm);
  const asteriskBullets = content.match(/^[\t ]*\*[\t ][^\n]+/gm);
  const plusBullets = content.match(/^[\t ]*\+[\t ][^\n]+/gm);

  if (hyphenBullets?.length) result.bullet = "-";
  else if (asteriskBullets?.length) result.bullet = "*";
  else if (plusBullets?.length) result.bullet = "+";

  // Ordered list marker detection
  const dotOrdered = content.match(/^\s*\d+\.[\t ][^\n]+/gm);
  const parenthesisOrdered = content.match(/^\s*\d+\)[\t ][^\n]+/gm);

  if (dotOrdered?.length) result.bulletOrdered = ".";
  else if (parenthesisOrdered?.length) result.bulletOrdered = ")";

  // Emphasis style detection
  const asteriskEmphasis = content.match(/(?<!\*)\*[^*\n]+\*(?!\*)/g);
  const underscoreEmphasis = content.match(/(?<!_)_[^_\n]+_(?!_)/g);

  if (asteriskEmphasis?.length) result.emphasis = "*";
  else if (underscoreEmphasis?.length) result.emphasis = "_";

  // Strong emphasis detection
  const asteriskStrong = content.match(/\*\*[^*\n]+\*\*/g);
  const underscoreStrong = content.match(/__[^_\n]+__/g);

  if (asteriskStrong?.length) result.strong = "*";
  else if (underscoreStrong?.length) result.strong = "_";

  // Code fence detection
  const backtickFence = content.match(/^```[^`]*```/gm);
  const tildeFence = content.match(/^~~~[^~]*~~~/gm);

  if (backtickFence?.length) result.fence = "`";
  else if (tildeFence?.length) result.fence = "~";

  // List item indentation detection
  const listItems = content.match(/^[\t ]*([-+*]|\d+[.)])\s+[^\n]+/gm);
  if (listItems?.length) {
    const indentTypes = new Set(
      listItems.map((item) => {
        const leadingSpace = item.match(/^[\t ]*/)?.[0] || "";
        if (leadingSpace.includes("\t")) return "tab";
        return leadingSpace.length === 1 ? "one" : leadingSpace.length > 1 ? "mixed" : "one";
      }),
    );

    if (indentTypes.size === 1) {
      result.listItemIndent = indentTypes.values().next().value;
    } else {
      result.listItemIndent = "mixed";
    }
  }

  // Horizontal rule detection
  const asteriskRule = content.match(/^[\t ]*(\*[\t ]*){3,}$/m);
  const hyphenRule = content.match(/^[\t ]*(-[\t ]*){3,}$/m);
  const underscoreRule = content.match(/^[\t ]*(_[\t ]*){3,}$/m);

  if (asteriskRule?.length) result.rule = "*";
  else if (hyphenRule?.length) result.rule = "-";
  else if (underscoreRule?.length) result.rule = "_";

  return result;
}
