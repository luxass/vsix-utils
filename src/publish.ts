/**
 * @module publish
 *
 * This module contains utility functions for publishing vsix files.
 */

import type { Manifest } from "./types";
import type { VSCE_TARGETS } from "./vsce-constants";

export interface PublishOptions {
  personalAccessToken?: string;

  targets?: typeof VSCE_TARGETS[];

  packagePaths?: string[];

  /**
   * Additional headers to apply to the request
   */
  additonalHeaders?: Record<string, string>;

  /**
   * Which registry to publish extensions to
   * @default "vsmarketplace"
   *
   * NOTE:
   * You can pass a object with registryUrl, if you are using a custom registry.
   */
  registry?: "vsmarketplace" | "openvsx" | { registryUrl: string };
}

export async function publish(options: PublishOptions): Promise<any> {
  const { registry = "vsmarketplace", additonalHeaders, packagePaths, targets, personalAccessToken } = options;

  const registryUrl = (typeof registry === "object" && "registryUrl" in registry) ? registry.registryUrl : registry === "openvsx" ? "https://open-vsx.org" : "https://marketplace.visualstudio.com";

  const publishRequests = [];
  // support multiple targets combined with multiple package paths
  // support multiple package paths without targets (auto detect target)
}
