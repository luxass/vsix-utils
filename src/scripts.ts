import type { Manifest, PackageManager } from "./types";
import { spawn } from "node:child_process";
import { promisify } from "node:util";

const spawnAsync = promisify(spawn);

export interface PrepublishOptions {
  /**
   * The current working directory.
   */
  cwd: string;

  /**
   * The package manager to use.
   */
  packageManager: Exclude<PackageManager, "auto">;

  /**
   * The manifest of the extension.
   */
  manifest: Manifest;

  /**
   * Whether the extension is a pre-release.
   */
  preRelease: boolean;
}

export async function prepublish(options: PrepublishOptions): Promise<void> {
  const { cwd, packageManager, manifest, preRelease } = options;

  if (!manifest.scripts) {
    return;
  }

  const scripts: string[] = [];
  if ("vscode:prepublish" in manifest.scripts) {
    scripts.push("vscode:prepublish");
  }

  if ("vscode:prepublish:release" in manifest.scripts && !preRelease) {
    scripts.push("vscode:prepublish:release");
  }

  if ("vscode:prepublish:prerelease" in manifest.scripts && preRelease) {
    scripts.push("vscode:prepublish:prerelease");
  }

  if (scripts.length === 0) {
    return;
  }

  for (const script of scripts) {
    await runScript({ cwd, packageManager, script });
  }
}

export interface RunScriptOptions {
  /**
   * The current working directory.
   */
  cwd: string;

  /**
   * The package manager to use.
   */
  packageManager: Exclude<PackageManager, "auto">;

  /**
   * The script to run.
   */
  script: string;
}

export async function runScript(options: RunScriptOptions): Promise<void> {
  const { cwd, packageManager, script } = options;

  await spawnAsync(packageManager, ["run", script], {
    cwd,
    shell: true,
    stdio: "inherit",
  });
}
