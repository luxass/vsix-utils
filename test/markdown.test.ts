import { readFile } from "node:fs/promises";
import { describe, expect, it } from "vitest";
import { detectMarkdownSyntax, inferBaseUrls, transformMarkdown } from "../src/markdown";
import { createManifest } from "./utils";

describe("transformMarkdown", () => {
  it("transform relative links", async () => {
    const content = await readFile("./test/fixtures/markdown/README.md", "utf-8");

    expect(content).toBeDefined();

    const manifest = createManifest({
      repository: "https://github.com/luxass/vsix-utils.git",
    });

    const transformed = await transformMarkdown(manifest, {
      content,
    });

    expect(transformed).toMatchSnapshot();
  });

  it("transform relative links with custom base URLs", async () => {
    const content = await readFile("./test/fixtures/markdown/README.md", "utf-8");

    expect(content).toBeDefined();

    const manifest = createManifest({
      repository: "https://github.com/luxass/vsix-utils.git",
    });

    const transformed = await transformMarkdown(manifest, {
      content,
      baseContentUrl: "https://github.com/luxass/luxass.dev",
      baseImagesUrl: "https://github.com/luxass/luxass.dev",
    });

    expect(transformed).toMatchSnapshot();
  });

  it("should throw an error if the repository is not detected", async () => {
    const content = await readFile("./test/fixtures/markdown/README.md", "utf-8");

    expect(content).toBeDefined();

    const manifest = createManifest();

    await expect(() =>
      transformMarkdown(manifest, {
        content,
      }),
    ).rejects.toThrowError("Couldn't detect the repository where this extension is published.");
  });
});

describe("inferBaseUrls", () => {
  it("should handle GitHub HTTPS URLs", () => {
    const manifest = createManifest({
      repository: "https://github.com/luxass/vsix-utils.git",
    });

    const urls = inferBaseUrls(manifest, "HEAD");

    expect(urls).toEqual({
      contentUrl: "https://github.com/luxass/vsix-utils/blob/HEAD",
      imagesUrl: "https://github.com/luxass/vsix-utils/raw/HEAD",
    });
  });

  it("should handle GitHub SSH URLs", () => {
    const manifest = createManifest({
      repository: "git@github.com:luxass/vsix-utils.git",
    });

    const urls = inferBaseUrls(manifest, "HEAD");

    expect(urls).toEqual({
      contentUrl: "https://github.com/luxass/vsix-utils/blob/HEAD",
      imagesUrl: "https://github.com/luxass/vsix-utils/raw/HEAD",
    });
  });

  it("should handle GitLab URLs", () => {
    const manifest = createManifest({
      repository: "https://gitlab.com/luxass/vsix-utils.git",
    });

    const urls = inferBaseUrls(manifest, "HEAD");

    expect(urls).toEqual({
      contentUrl: "https://gitlab.com/luxass/vsix-utils/-/blob/HEAD",
      imagesUrl: "https://gitlab.com/luxass/vsix-utils/-/raw/HEAD",
    });
  });

  it("should handle repository object format", () => {
    const manifest = createManifest({
      repository: {
        type: "git",
        url: "https://github.com/luxass/vsix-utils.git",
      },
    });

    const urls = inferBaseUrls(manifest, "HEAD");

    expect(urls).toEqual({
      contentUrl: "https://github.com/luxass/vsix-utils/blob/HEAD",
      imagesUrl: "https://github.com/luxass/vsix-utils/raw/HEAD",
    });
  });

  it("should return null for invalid repository", () => {
    const manifest = createManifest({
      repository: "invalid-url",
    });

    const urls = inferBaseUrls(manifest, "HEAD");

    expect(urls).toBeNull();
  });
});

describe("detectMarkdownSyntax", () => {
  it("should detect unordered list bullet style", () => {
    expect(detectMarkdownSyntax("* List item")).toEqual({ bullet: "*", listItemIndent: "one" });
    expect(detectMarkdownSyntax("- List item")).toEqual({ bullet: "-", listItemIndent: "one" });
    expect(detectMarkdownSyntax("+ List item")).toEqual({ bullet: "+", listItemIndent: "one" });
  });

  it("should detect ordered list marker style", () => {
    expect(detectMarkdownSyntax("1. First item")).toEqual({ bulletOrdered: ".", listItemIndent: "one" });
    expect(detectMarkdownSyntax("1) First item")).toEqual({ bulletOrdered: ")", listItemIndent: "one" });
  });

  it("should detect emphasis style", () => {
    expect(detectMarkdownSyntax("*italic*")).toEqual({ emphasis: "*" });
    expect(detectMarkdownSyntax("_italic_")).toEqual({ emphasis: "_" });
  });

  it("should detect strong emphasis style", () => {
    expect(detectMarkdownSyntax("**bold**")).toEqual({ strong: "*" });
    expect(detectMarkdownSyntax("__bold__")).toEqual({ strong: "_" });
  });

  it("should detect code fence style", () => {
    expect(detectMarkdownSyntax("```code```")).toEqual({ fence: "`" });
    expect(detectMarkdownSyntax("~~~code~~~")).toEqual({ fence: "~" });
  });

  it("should detect list item indentation", () => {
    expect(detectMarkdownSyntax("  * List")).toEqual({ bullet: "*", listItemIndent: "mixed" });
    expect(detectMarkdownSyntax("\t* List")).toEqual({ bullet: "*", listItemIndent: "tab" });
    expect(detectMarkdownSyntax("   * List")).toEqual({ bullet: "*", listItemIndent: "mixed" });
  });

  it("should detect horizontal rule style", () => {
    expect(detectMarkdownSyntax("***")).toEqual({ rule: "*" });
    expect(detectMarkdownSyntax("---")).toEqual({ rule: "-" });
    expect(detectMarkdownSyntax("___")).toEqual({ rule: "_" });
  });

  it("should detect multiple syntax elements", () => {
    const content = `
* List item
**bold text**
---
`;
    expect(detectMarkdownSyntax(content)).toEqual({
      bullet: "*",
      strong: "*",
      rule: "-",
      listItemIndent: "one",
    });
  });

  it("should return empty object for no matches", () => {
    expect(detectMarkdownSyntax("Plain text")).toEqual({});
  });
});
