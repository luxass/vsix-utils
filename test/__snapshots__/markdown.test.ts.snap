// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transformMarkdown > transform relative links 1`] = `
"# Project Documentation

## Getting Started

Please review our [installation guide](https://github.com/luxass/vsix-utils/blob/HEAD/docs/installation.md) before proceeding to the [quick start tutorial](https://github.com/luxass/vsix-utils/blob/HEAD/docs/quickstart.md).

Visit our repositories on [GitHub (HTTPS)](https://github.com/example/project) or [GitHub (HTTP)](http://github.com/example/project-mirror).

Check out our [partner projects](https://github.com/luxass/vsix-utils/blob/HEAD/partners/index.md).

## Project Structure

\`\`\`
project/
├── docs/
│   ├── installation.md
│   └── quickstart.md
├── assets/
│   ├── images/
│   │   └── logo.png
│   └── videos/
│       └── demo.mp4
└── src/
    └── components/
\`\`\`

## Documentation

Our documentation is split into several sections:

- [API Reference](https://github.com/luxass/vsix-utils/blob/HEAD/docs/api/index.md)
- [Component Library](https://github.com/luxass/vsix-utils/blob/HEAD/src/components/README.md)
- [Contributing Guidelines](https://github.com/luxass/vsix-utils/blob/HEAD/CONTRIBUTING.md)
- [Code of Conduct](https://example.com/code-of-conduct)

## Media Examples

### Images

Here's our project logo:
![Project Logo](https://github.com/luxass/vsix-utils/raw/HEAD/assets/images/logo.png)

Some examples using HTML img tags:

<img src="https://github.com/luxass/vsix-utils/raw/HEAD/assets/images/screenshot1.png" alt="Screenshot 1" width="600">
<img src="https://github.com/luxass/vsix-utils/raw/HEAD/assets/images/banner.jpg" alt="Banner" width="800">
<img src="https://example.com/images/team.jpg" alt="Team Photo">

### Videos

Project demo video:
![Demo Video](https://github.com/luxass/vsix-utils/raw/HEAD/assets/videos/demo.mp4)

Using HTML video tags:

<video width="720" height="480" controls>
  <source src="https://github.com/luxass/vsix-utils/blob/HEAD/assets/videos/tutorial.mp4" type="video/mp4">
  <source src="https://github.com/luxass/vsix-utils/blob/HEAD/assets/videos/tutorial.webm" type="video/webm">
  Your browser does not support the video tag.
</video>

## External Resources

- [Main Repository](https://github.com/luxass/vsix-utils/blob/HEAD/README.md)
- [Documentation Site](https://docs.example.com)
- [Issue Tracker](https://github.com/luxass/vsix-utils/blob/HEAD/issues/index.md)
- [Team Calendar](https://github.com/luxass/vsix-utils/blob/HEAD/team/calendar)

## Navigation

- [Back to Home](https://github.com/luxass/vsix-utils/blob/HEAD/README.md)
- [Go to Dashboard](https://github.com/luxass/vsix-utils/blob/HEAD/../dashboard/index.md)
- [View Settings](https://github.com/luxass/vsix-utils/blob/HEAD/settings/index.md)

## Component Examples

- [Button Component](https://github.com/luxass/vsix-utils/blob/HEAD/src/components/Button/README.md)
- [Form Elements](https://github.com/luxass/vsix-utils/blob/HEAD/src/components/Form/index.md)
- [Navigation Bar](https://github.com/luxass/vsix-utils/blob/HEAD/src/components/NavBar/documentation.md)

## Additional Resources

For more information, check out these resources:

1. [Developer Blog](https://github.com/luxass/vsix-utils/blob/HEAD/blog/index.md)
2. [Release Notes](https://github.com/luxass/vsix-utils/blob/HEAD/changelog.md)
3. [Security Policy](https://github.com/luxass/vsix-utils/blob/HEAD/SECURITY.md)
4. [License Information](https://github.com/luxass/vsix-utils/blob/HEAD/LICENSE)

---

*Last updated: 2025-01-14*

[Return to top](#project-documentation)
"
`;

exports[`transformMarkdown > transform relative links with custom base URLs 1`] = `
"# Project Documentation

## Getting Started

Please review our [installation guide](https://github.com/luxass/luxass.dev/docs/installation.md) before proceeding to the [quick start tutorial](https://github.com/luxass/luxass.dev/docs/quickstart.md).

Visit our repositories on [GitHub (HTTPS)](https://github.com/example/project) or [GitHub (HTTP)](http://github.com/example/project-mirror).

Check out our [partner projects](https://github.com/luxass/luxass.dev/partners/index.md).

## Project Structure

\`\`\`
project/
├── docs/
│   ├── installation.md
│   └── quickstart.md
├── assets/
│   ├── images/
│   │   └── logo.png
│   └── videos/
│       └── demo.mp4
└── src/
    └── components/
\`\`\`

## Documentation

Our documentation is split into several sections:

- [API Reference](https://github.com/luxass/luxass.dev/docs/api/index.md)
- [Component Library](https://github.com/luxass/luxass.dev/src/components/README.md)
- [Contributing Guidelines](https://github.com/luxass/luxass.dev/CONTRIBUTING.md)
- [Code of Conduct](https://example.com/code-of-conduct)

## Media Examples

### Images

Here's our project logo:
![Project Logo](https://github.com/luxass/luxass.dev/assets/images/logo.png)

Some examples using HTML img tags:

<img src="https://github.com/luxass/luxass.dev/assets/images/screenshot1.png" alt="Screenshot 1" width="600">
<img src="https://github.com/luxass/luxass.dev/assets/images/banner.jpg" alt="Banner" width="800">
<img src="https://example.com/images/team.jpg" alt="Team Photo">

### Videos

Project demo video:
![Demo Video](https://github.com/luxass/luxass.dev/assets/videos/demo.mp4)

Using HTML video tags:

<video width="720" height="480" controls>
  <source src="https://github.com/luxass/luxass.dev/assets/videos/tutorial.mp4" type="video/mp4">
  <source src="https://github.com/luxass/luxass.dev/assets/videos/tutorial.webm" type="video/webm">
  Your browser does not support the video tag.
</video>

## External Resources

- [Main Repository](https://github.com/luxass/luxass.dev/README.md)
- [Documentation Site](https://docs.example.com)
- [Issue Tracker](https://github.com/luxass/luxass.dev/issues/index.md)
- [Team Calendar](https://github.com/luxass/luxass.dev/team/calendar)

## Navigation

- [Back to Home](https://github.com/luxass/luxass.dev/README.md)
- [Go to Dashboard](https://github.com/luxass/luxass.dev/../dashboard/index.md)
- [View Settings](https://github.com/luxass/luxass.dev/settings/index.md)

## Component Examples

- [Button Component](https://github.com/luxass/luxass.dev/src/components/Button/README.md)
- [Form Elements](https://github.com/luxass/luxass.dev/src/components/Form/index.md)
- [Navigation Bar](https://github.com/luxass/luxass.dev/src/components/NavBar/documentation.md)

## Additional Resources

For more information, check out these resources:

1. [Developer Blog](https://github.com/luxass/luxass.dev/blog/index.md)
2. [Release Notes](https://github.com/luxass/luxass.dev/changelog.md)
3. [Security Policy](https://github.com/luxass/luxass.dev/SECURITY.md)
4. [License Information](https://github.com/luxass/luxass.dev/LICENSE)

---

*Last updated: 2025-01-14*

[Return to top](#project-documentation)
"
`;
